bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java

```diff
@@ -0,0 +1,169 @@
+
+    /**
+     * constructor
+     *
+     * @param executor the executor to use
+     * @param errorHandler the errorHandler
+     * @param pingPeriod the period pings should be sent
+     */
+    WebSocketConnectionImpl(ScheduledExecutorService executor, Consumer<String> errorHandler, Duration pingPeriod) {
+        this.errorHandler = errorHandler;
+        this.pingPeriod = pingPeriod;
+        this.pingHandle = executor.scheduleAtFixedRate(this::sendPing, pingPeriod.dividedBy(2).toMillis(),
+                pingPeriod.toMillis(), TimeUnit.MILLISECONDS);
+        this.watchDogHandle = executor.scheduleWithFixedDelay(this::checkConnection, pingPeriod.toMillis(),
+                pingPeriod.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() {
+        var session = sessionRef.getAndSet(null);
+        if (session != null) {
+            session.close();
+        }
+        watchDogHandle.cancel(true);
+        pingHandle.cancel(true);
+    }
+
+    @Override
+    public void observeProperty(@NotNull String propertyName, @NotNull BiConsumer<String, Object> listener) {
+        propertyChangedListeners.put(propertyName, listener);
+    }
+
+    @Override
+    public void onWebSocketConnect(Session session) {
+        sessionRef.set(session); // save websocket session to be able to send ping
+    }
+
+    @Override
+    public void onWebSocketPing(ByteBuffer payload) {
+    }
+
+    @Override
+    public void onWebSocketPong(ByteBuffer payload) {
+        lastTimeReceived.set(Instant.now());
+    }
+
+    @Override
+    public void onWebSocketBinary(byte[] payload, int offset, int len) {
+    }
+
+    @Override
+    public void onWebSocketText(String message) {
+        try {
+            var propertyStatus = new Gson().fromJson(message, PropertyStatusMessage.class);
+            if (propertyStatus.messageType.equals("propertyStatus")) {
+                for (var propertyEntry : propertyStatus.data.entrySet()) {
+                    propertyChangedListeners.getOrDefault(propertyEntry.getKey(), EMPTY_PROPERTY_CHANGED_LISTENER)
+                            .accept(propertyEntry.getKey(), propertyEntry.getValue());
+                }
+            } else {
+                logger.debug("Ignoring received message of unknown type: {}", message);
+            }
+        } catch (JsonSyntaxException se) {
+            logger.warn("received invalid message: {}", message);
+        }
+    }
+
+    @Override
+    public void onWebSocketClose(int statusCode, String reason) {
+        onWebSocketError(new IOException("websocket closed by peer. " + Optional.ofNullable(reason).orElse("")));
+    }
+
+    @Override
+    public void onWebSocketError(Throwable cause) {
+        onError(cause.getMessage());
+    }
+
+    private void onError(String message) {
+        try {
+            errorHandler.accept(message);
+        } catch (Exception e) {
```

Comment:
What type of exception do you expect here? Can you specify the concrete type?

Reply:
Can you comment on my question?

Reply:
catching exception removed